### PR TITLE
JAMES-3754 Searching support IMAP savedate extension

### PR DESCRIPTION
**Pull request by quantranhong1999**
TODO:
- [x] Lucene implementation

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 1]
```diff
@@ -1278,6 +1296,50 @@ public String toString() {
...
 
+    /**
+     * Filters on the save date.
+     */
+    public static class SaveDateCriterion extends Criterion {
+
+        private final DateOperator operator;
+
+        public SaveDateCriterion(DateOperator operator) {
+            super();
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java:1307]: Not needed?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 2]
```diff
@@ -1278,6 +1296,50 @@ public String toString() {
...
+         * Gets the search operation and value to be evaluated.
+         *
+         * @return the <code>Operator</code>, not null
+         */
+        public DateOperator getOperator() {
+            return operator;
+        }
+
+        @Override
+        public int hashCode() {
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java:1321]:
```suggestion
        public final int hashCode() {
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 3]
```diff
@@ -1278,6 +1296,50 @@ public String toString() {
...
+            return Objects.hashCode(operator);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java:1326]:
```suggestion
        public final boolean equals(Object obj) {
```

**Review by chibenwa — COMMENTED** [review]
Idem, looks good so far.

**Comment by Arsnael** [comment]
Conflicts in the branch

**Review by Arsnael — APPROVED** [review]
Read it, looks good to me as well

**Comment by Arsnael** [comment]
Conflicts, rebase please

File: mailbox/cassandra/pom.xml [hunk 4]
```diff
@@ -129,6 +129,11 @@
             <artifactId>james-server-task-memory</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-testing</artifactId>
+            <scope>runtime</scope>
```
Comment (vttranlina) [inline mailbox/cassandra/pom.xml:135]: scope test is better?

File: mailbox/jpa/pom.xml [hunk 5]
```diff
@@ -89,6 +89,11 @@
             <artifactId>james-server-data-jpa</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-testing</artifactId>
+            <scope>runtime</scope>
```
Comment (vttranlina) [inline mailbox/jpa/pom.xml:95]: scope test is better?

File: mailbox/cassandra/pom.xml (hunk 4 above)
Comment (quantranhong1999) [reply mailbox/cassandra/pom.xml:135]: The maven test module `james-server-testing` with scope `test` is not transitive to other testing modules (lead to NoDefClassFound..). 

So I use `runtime` scope: `This scope indicates that the dependency is not required for compilation, but is for execution. Maven includes a dependency with this scope in the runtime and test classpaths, but not the compile classpath.`
rf: https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope

Better idea?

**Comment by quantranhong1999** [comment]
Can we merge this?

**Pull request merged**